src/main.rs

```diff
@@ -0,0 +1,188 @@
+use std::error::Error;
+use std::ffi::{CStr, CString};
+use std::fmt::{Display, Formatter};
+use winit::event_loop::EventLoop;
+use winit::raw_window_handle::HasWindowHandle;
+use winit::window::{Window, WindowAttributes};
+
+use glutin::config::{Config, ConfigTemplateBuilder};
+use glutin::context::{ContextApi, ContextAttributesBuilder, NotCurrentContext, Version};
+use glutin::display::GetGlDisplay;
+use glutin::prelude::*;
+
+use glutin_winit::{DisplayBuilder, GlWindow};
+
+pub mod gl {
+    #![allow(clippy::all)]
+    include!(concat!(env!("OUT_DIR"), "/gl_bindings.rs"));
+
+    pub use Gles2 as Gl;
+}
+
+#[derive(Debug, Clone)]
+struct GLInfo {
+    driver: String,
+    vendor: String,
+    renderer: String,
+    version: String,
+    shading_language: String,
+}
+
+impl Display for GLInfo {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.write_str(&format!(
+            "{} Vendor: {}\n\
+             Renderer: {}\n\
+             Version: {}\n\
+             Shading Language: {}",
+            self.driver, self.vendor, self.renderer, self.version, self.shading_language
+        ))
+    }
+}
+
+pub fn main() -> Result<(), Box<dyn Error>> {
+    let args: Vec<String> = std::env::args().collect();
+    if args.len() == 2 && (args[1] == "-h" || args[1] == "--help") {
+        println!("Usage: glinfo [-f filename]");
+        return Ok(());
+    }
+
+    let event_loop =
+        EventLoop::new().map_err(|err| format!("Failed to create event loop: {}", err))?;
+    let template = ConfigTemplateBuilder::new();
+
+    let display_builder = DisplayBuilder::new()
+        .with_window_attributes(Some(Window::default_attributes().with_visible(false)));
+
+    let output = match get_gl_info(template, display_builder, &event_loop) {
+        Ok(gl_info) => format!("{}", gl_info),
+        Err(err) => format!("ERROR: {}", err),
+    };
+
+    println!("{}", output);
+
+    if args.len() == 3 && args[1] == "-f" {
+        let filename = &args[2];
+        std::fs::write(filename, output.as_bytes())?;
+    }
+
+    Ok(())
+}
+
+fn get_gl_info(
+    template: ConfigTemplateBuilder,
+    display_builder: DisplayBuilder,
+    event_loop: &EventLoop<()>,
+) -> Result<GLInfo, String> {
+    // We just created the event loop, so initialize the display, pick the config, and
+    // create the context.
+    let (window, gl_config) =
+        match display_builder
+            .clone()
```

Comment:
```suggestion
```
I believe that since this function takes an owned `display_builder`, you can safely consume it in `.build(...)`.

Alternatively you can change the function to receive a borrowed `DisplayBuilder` (by reference).

The same applies to `template`.

Of course here we're not worried about the performance overhead of cloning these two, but the calls to `.clone()` are just noise in the code.